sale_order_price_recalculation/README.rst

```diff
@@ -0,0 +1,113 @@
+===================================
+Price recalculation in sales orders
+===================================
+
+.. !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+   !! This file is generated by oca-gen-addon-readme !!
+   !! changes will be overwritten.                   !!
+   !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+.. |badge1| image:: https://img.shields.io/badge/maturity-Beta-yellow.png
+    :target: https://odoo-community.org/page/development-status
+    :alt: Beta
+.. |badge2| image:: https://img.shields.io/badge/licence-AGPL--3-blue.png
+    :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+    :alt: License: AGPL-3
+.. |badge3| image:: https://img.shields.io/badge/github-OCA%2Fsale--workflow-lightgray.png?logo=github
+    :target: https://github.com/OCA/sale-workflow/tree/12.0/sale_order_price_recalculation
+    :alt: OCA/sale-workflow
+.. |badge4| image:: https://img.shields.io/badge/weblate-Translate%20me-F47D42.png
+    :target: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_order_price_recalculation
+    :alt: Translate me on Weblate
+.. |badge5| image:: https://img.shields.io/badge/runbot-Try%20me-875A7B.png
+    :target: https://runbot.odoo-community.org/runbot/167/12.0
+    :alt: Try me on Runbot
+
+|badge1| |badge2| |badge3| |badge4| |badge5| 
+
+This module add 2 buttons on sale orders (below sale order lines) that:
+
+* Recalculates the prices of the order lines that contain a product in them.
+* Reset product descriptions from current product information.
+
+It is launched manually as a button to get the user decide if he/she wants to
+recalculate prices when pricelist is changed or after duplicating a sale order
+to update or not sales information.
+
+**Table of contents**
+
+.. contents::
+   :local:
+
+Installation
+============
+
+You need to install sale_management module for accessing the needed menus.
+
+Usage
+=====
+
+Inside a sale order, you can click on "Recalculate prices" to launch a
+recalculation of all the prices of the lines, losing previous custom prices.
+
+The second "Reset descriptions" will get descriptions from products, losing
+custom descriptions.
+
+.. image:: /sale_order_price_recalculation/static/description/sale_order_price_recalculation.png
+    :alt: Sale order price recalculation
+
+Known issues / Roadmap
+======================
+
+* In a sale order with lot of lines, the recalculation may slow down, because
+  sale general data (amount untaxed, amount taxed...) are recalculated for
+  each line.
```

Comment:
You might have a look at the two methods here to boost performance:
- disable prefetch of fields
- delay recompute until all changes that affect the computation have been done, that is in this case, until after all lines have been updated: https://github.com/OCA/contract/blob/f9c3e98a274610bb21be1b81f81c940b0ed131d1/contract/models/account_analytic_account.py#L325

Reply:
Good ideas!

Reply:
`    @api.multi
    def recalculate_prices(self):
        _self = self.with_context(prefetch_fields=False)
        with _self.env.norecompute():
            for line in _self.mapped('order_line'):
                dict = line._convert_to_write(line.read()[0])
                if 'product_tmpl_id' in line._fields:
                    dict['product_tmpl_id'] = line.product_tmpl_id
                line2 = _self.env['sale.order.line'].new(dict)
                # we make this to isolate changed values:
                line2.product_uom_change()
                line.price_unit = line2.price_unit
        _self.recompute()
        return True

    @api.multi
    def recalculate_names(self):
        _self = self.with_context(prefetch_fields=False)
        with _self.env.norecompute():
            for line in _self.mapped('order_line').filtered('product_id'):
                # we make this to isolate changed values:
                line2 = _self.env['sale.order.line'].new({
                    'product_id': line.product_id,
                })
                line2.product_id_change()
                line.name = line2.name
        _self.recompute()
        return True`

Like this?
Do I have to recompute the orde lines also?

Reply:
I think this should be enough, as you already update the move lines with the information you want. But to be totally sure you could test this, or even better add a unit test.

Reply:
@rven are you going to apply these improvement?